pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -191,41 +195,50 @@ def restore_loops(self) -> None:
-        expected_steps = self.trainer.num_training_batches / n_accum
-        if self.trainer.num_training_batches != 0 and self.trainer.global_step % expected_steps > 1:
-            rank_zero_warn(
-                "You're resuming from a checkpoint that ended mid-epoch."
-                " Training will start from the beginning of the next epoch."
-                " This can cause unreliable results if further training is done,"
-                " consider using an end of epoch checkpoint."
+        if self.trainer.state.fn == TrainerFn.FITTING:
+            self.trainer.fit_loop.global_step = self._loaded_checkpoint["global_step"]
+            self.trainer.fit_loop.current_epoch = self._loaded_checkpoint["epoch"]
+
+            # crash if max_epochs is lower then the current epoch from the checkpoint
+            if (
+                FitLoop._is_max_limit_enabled(self.trainer.max_epochs)
+                and self.trainer.max_epochs is not None
+                and self.trainer.current_epoch > self.trainer.max_epochs
+            ):
+                raise MisconfigurationException(
+                    f"You restored a checkpoint with current_epoch={self.trainer.current_epoch},"
+                    f" but you have set Trainer(max_epochs={self.trainer.max_epochs})."
+                )
+
+            # Division deals with global step stepping once per accumulated batch
+            # Inequality deals with different global step for odd vs even num_training_batches
+            self.trainer.accumulate_grad_batches = self.trainer.accumulation_scheduler.get_accumulate_grad_batches(
+                self.trainer.current_epoch
             )
+            n_accum = 1 if self.trainer.accumulate_grad_batches is None else self.trainer.accumulate_grad_batches
+            expected_steps = self.trainer.num_training_batches / n_accum
+            if self.trainer.num_training_batches != 0 and self.trainer.global_step % expected_steps > 1:
+                rank_zero_warn(
+                    "You're resuming from a checkpoint that ended mid-epoch."
+                    " Training will start from the beginning of the next epoch."
+                    " This can cause unreliable results if further training is done,"
+                    " consider using an end of epoch checkpoint."
+                )
 
         state_dict = self._loaded_checkpoint.get("loops")
-        if state_dict:
-            self.trainer.fit_loop.load_state_dict(state_dict["fit_loop"])
-            self.trainer.validate_loop.load_state_dict(state_dict["validate_loop"])
-            self.trainer.test_loop.load_state_dict(state_dict["test_loop"])
-            self.trainer.predict_loop.load_state_dict(state_dict["predict_loop"])
+
+        if state_dict is not None:
+            if self.trainer.state.fn == TrainerFn.FITTING:
+                self.trainer.fit_loop.load_state_dict(state_dict["fit_loop"])
+
+            if self.trainer.state.fn == TrainerFn.VALIDATING:
```

Comment:
This means that before you would only need to load once, but now you need to do it per trainer fn if you are using a checkpoint with existing data per loop.

Not sure if this is a good thing.

Can you add a test?

Reply:
I think we have the reload for val_loop inside fitting too since evaluation is also a part for `trainer.fit`. For others, I guess it's fine since others are isolated, and no matter what this will be called everytime with `trainer.{test/validate/predict}`

Reply:
even before, it was called every time you call `trainer.{fit/test/validate/predict}`, but now let's say if `trainer.test` is not called then there is no point in loading the `test_loop` state. I am not sure about `val_loop`, whether to load it or not inside FITTING, since for a normal case, evaluation with `trainer.fit` goes over a complete epoch from scratch. do you think loading `val_loop` state dict inside FITTING makes sense?>